### DISCUSSION TITLE
Universal MIDI Packet support in JACK

### DISCUSSION THREAD

**Discussion by bbouchez** [jackaudio/jack2#985] [Show and tell]
Hello all,

I have released the source code of the two demos I made to show that JACK was natively compatible with the new Universal MIDI Packet format (UMP) from the MMA.

https://github.com/bbouchez/jackumptestmonitor
https://github.com/bbouchez/jacktestumpsender

As JACK has no idea of what UMP is, the related streams appear as sequence of bytes which must be interpreted by the JACK client. The "monitor" application shows how to rebuild the UMP words (as UMP is based only on 32-bit packets, not bytes)
In theory, MIDI-CI should also be used when a UMP application wants to send UMP packets over JACK (as the other application may be compatible only with native MIDI 1.0 protocol)

I have started to write some utilities / helpers functions to help sending / decoding UMP between JACK applications.

Benoit

**Comment by jcelerier** [comment]
hey that's great ! I'm curious about the discrimination between non-ump and ump packet if we don't know: how can one differentiate between

    0xF0 0x00 0x00 [...12 other or so...] 0xF7
    
which could be either some random MIDI 1 sysex, or some MIDI 2 Endpoint Discovery Message Format if I'm not mistaken ?

e.g. this : 

![image](https://github.com/user-attachments/assets/0e93bb5a-3b3b-43ed-bf63-e23f08bdbf02)

**Reply by bbouchez** [reply]
Hello Jean-Michael,

in fact, it is not possible to make a 100% reliable discrimination. In my code, I do that *very* roughly : I consider that any packet which is 4, 8, 12 or 16 bytes long is UMP. So yes, there is a risk that a MIDI 1.0 SYSEX is accidentally considered as a MT=0xF valid message (problem is that you can find valid UMP messages starting with 0xF0 and having the last word being something 0xxxxxxxF7)

For now, what I do is simply a manual setup (I make sure that the sender and receiver I connect through JACK both support UMP - which is not a difficult task for now, as the UMP applications I use are the ones I have written °-)

In the future, the best thing would be to implement Endpoint Discovery Message and send it when JACK informs the application it has been connected to something : this message can not be mismatched with a MIDI 1.0 as it starts with F0 but ends with 00 (it is not possible to have a F7 in the last byte). If the receiver does not reply with Device Identity message, it would mean it does support UMP
This would however make the implementation of these messages mandatory in all JACK applications which would support UMP, which I don't especially like

**Comment by symdeb** [comment]
I guess the best way would be for JACK to use the ALSA API to get the port properties so that clients can get the UMP protocol, jitter flags, sysex8 streams and the group used for the ports (from JACK) , Probably that would apply for the pipewire jack emulation as well.
